### ISSUE TITLE
500 Missing Scope error when using incoming-webhook bot token scope

### ISSUE THREAD

**Issue by sedwardsgt** [laravel/slack-notification-channel#92]
### Slack Notification Channel Version

3.2.0

### Laravel Version

11.9.2

### PHP Version

8.2

### Database Driver & Version

mongodb/laravel-mongodb v4.4

### Description

In converting my app from Laravel 10 to 11, I am converting from using [nathanheffley/laravel-slack-blocks](https://github.com/nathanheffley/laravel-slack-blocks) to this library. We. have a working published app that our users have installed and enabled that uses OAuth to get a bot token and `incoming-webhook` URL. and then sends notifications to the designated Slack channel.

Per the [Laravel docs](https://laravel.com/docs/11.x/notifications#notifying-external-slack-workspaces), I have obtained a token and channel name from the [oauth response](https://api.slack.com/messaging/webhooks) and stored them in the database, and then written my `routeNotificationForSlack` in my notifiable model like so:
```
    public function routeNotificationForSlack($notification): mixed
    {
        return SlackRoute::make($this->getAttribute('slackChannelName'), $this->getAttribute('slackBotToken'));
    }
```

However, when I use [SlackMessage](https://github.com/laravel/slack-notification-channel/blob/3.x/src/Slack/SlackMessage.php) to make the call, I get a 500 error saying
```
Slack API call failed with error [missing_scope].
```
Per the docs - and what has always worked - the only bot scope that needs to be defined is `incoming-webhook`.

The reason I'm filing this here is that the only thing that has changed is my Laravel library/code, not anything on the Slack side.

### Steps To Reproduce

1. Enable `incoming-webhook` scope for bot token in Slack app.
2. Obtain bot token and channel name from OAuth response, and store locally
3. Implement `routeNotificationForSlack` in notifiable model  per [docs](https://laravel.com/docs/11.x/notifications#notifying-external-slack-workspaces) using token and channel value
4. Send message using [Illuminate\Notifications\Slack\SlackMessage](https://github.com/laravel/slack-notification-channel/blob/3.x/src/Slack/SlackMessage.php) from `toSlack()` implementation in notification class.

### Update ###
Digging further into the code and response from this call
```
$response = $this->http->asJson()
    ->withToken($route->token)
    ->post('https://slack.com/api/chat.postMessage', $payload)
    ->throw();
```
in [Illuminate\Http\Client\Response->json()](https://github.com/illuminate/http/blob/master/Client/Response.php#L73-L83), I get the following array returned from `$response->json()`
```
decoded = [
    ok = false
    error = "missing_scope"
    needed = "chat:write:bot"
    provided="incoming-webhook"
    warning="missing_charset"
]
```
However, the issues are 
1. There is no `chat:write:bot` permission/scope available in the app
2. If I add it as a `scope` param in my OAuth URL, it overwrites the `incoming-webhooks` settings.

What's even more confusing is that with the `nathanheffley/laravel-slack-blocks` library, my notifiable model call looked like this
```
    public function routeNotificationForSlack($notification): ?string
    {
        return $this->getAttribute('slackWebhookUrl');
    }
```
and works fine. However, if I do the same thing with this tool , I get [this issue](https://github.com/laravel/slack-notification-channel/issues/68), which a) doesn't have a definite answer and b) was closed for some reason, even though it wasn't totally solved (the same is true for [this similar issue](https://github.com/laravel/slack-notification-channel/issues/85)). As noted above, this library is calling `https://slack.com/api/chat.postMessage`, which is what is giving me the `missing_scope` error.

Looking at this and the other linked issues, this really needs either some fixing or clarification.

**Comment by jbrooksuk** [comment]
@sedwardsgt I'm afraid I'm not going to have time to look at this for a couple of weeks.

Also, it looks like this is only part of the error response? Are you able to attach the complete response from the Slack API - which is typically a JSON object.

In the meantime, if someone else is able to look into this, I'd appreciate a PR.

**Comment by sedwardsgt** [comment]
With some more work, I am now able to send messages, with a couple of caveats. After a few re-authorizations, I started getting the `[not_in_channel]` error message, which indicated that my bot had to be invited to my designated channel. Once I went to my channel in my workspace and did `/invite @myappname`, I was able to successfully send messages. But again, this raises two issues.

First, with our current implementation with [nathanheffley/laravel-slack-blocks](https://github.com/nathanheffley/laravel-slack-blocks), we don't have to do any bot invites. As I pointed out above, our `routeNotificationForSlack()` implementation just returns the slack URL that we got from OAuth (e.g. `https://hooks.slack.com/services/whatever`), and that's all that's needed. We don't have to make a `conversations.invite` call to do it.

Second, along the lines of the first point, this is now the call that is made in [SlackChannel->send()](https://github.com/laravel/slack-notification-channel/blob/3.x/src/Slack/SlackChannel.php#L35-L61)
```
    public function send(mixed $notifiable, Notification $notification): ?Response
    {
        $route = $this->determineRoute($notifiable, $notification);

        $message = $notification->toSlack($notifiable);

        $payload = $this->buildJsonPayload($message, $route);

        if (! $payload['channel']) {
            throw new LogicException('Slack notification channel is not set.');
        }

        if (! $route->token) {
            throw new LogicException('Slack API authentication token is not set.');
        }

        $response = $this->http->asJson()
            ->withToken($route->token)
            ->post('https://slack.com/api/chat.postMessage', $payload)
            ->throw();

        if ($response->successful() && $response->json('ok') === false) {
            throw new RuntimeException('Slack API call failed with error ['.$response->json('error').'].');
        }

        return $response;
    }
```
So what it's now doing is using `routeNotificationForSlack()` hook to just get the token and channel, and then posting to `https://slack.com/api/chat.postMessage` .

I don't understand the reason for the difference (and now the extra steps).

**Comment by Ticket-Master** [comment]
i dont know if this is relevant to your problem but to get a grasp of basics of setting up, you can visit this article from Ralph. I found it useful and when i setup according to it. i fixed my problem
https://ralphjsmit.com/laravel-slack-webhook-notifications

**Comment by lengjunbin-trec** [comment]
meet same error. Is there any progress on this issue？
```
Array:6 [
    "ok" => false
    "error" => "missing_scope"
    "needed" => "chat:write:bot"
    "provided" => "incoming-webhook"
    "warning" => "missing_charset"
    "response_metadata" => array:1 [
      "warnings" => array:1 [
        0 => "missing_charset"
      ]
    ]
  ]

**Comment by bhanwarpsrathore** [comment]
I checked Slack docs for [chat.postMessge](https://api.slack.com/methods/chat.postMessage) and found that it really needs `chat:write:bot`. @sedwardsgt I don't think it is channels issue. 

![image](https://github.com/user-attachments/assets/afb7c5df-074e-4d77-8ccb-c165e6955391)

**Comment by jbrooksuk** [comment]
@bhanwarpsrathore are you suggesting that this isn't an issue and you need to add the `chat:write:bot` permission?

**Comment by bhanwarpsrathore** [comment]
@jbrooksuk According to the Slack docs, Yes. You need to add the permission.

**Comment by jbrooksuk** [comment]
Thanks! We should probably get the docs updated there, https://laravel.com/docs/11.x/notifications#slack-prerequisites. Would you like to make the PR, or shall I? :)

**Comment by blift** [comment]
Hey,

I believe the current documentation and scope validation logic are outdated in relation to Slack’s latest API changes.

According to Slack’s documentation, the chat:write scope now replaces both chat:write:user and chat:write:bot. In fact, chat:write:bot is no longer listed among the available scopes:

“For [Slack apps](https://api.slack.com/authentication/basics), chat:write replaces both [chat:write:user](https://api.slack.com/scopes/chat:write:user) and [chat:write:bot](https://api.slack.com/scopes/chat:write:bot). [Here’s a bit more detail on why perspectival scopes are disappearing.](https://api.slack.com/authentication/quickstart#perspective)”

However, when using only chat:write, I still get a 500 error. I suspect this might be due to the package still validating against the now-deprecated chat:write:bot scope.

Link to the new scope: https://api.slack.com/scopes/chat:write

**Comment by bhanwarpsrathore** [comment]
Hi @blift, I still see `chat:write:user` and `chat:write:bot` for user tokens in slack doc. Here is the screenshot

<img width="874" alt="Image" src="https://github.com/user-attachments/assets/ee155056-0205-4e55-a9bd-c449d7342eb3" />

**Comment by blift** [comment]
Hey @bhanwarpsrathore , thanks for checking - and sorry for the late reply! You’re right, it’s still listed in the docs, which is a bit strange. Can I ask if you’re able to select this scope in your Slack panel? On my end, I don’t see chat:write:bot in either the bot or user scopes.

**Comment by bhanwarpsrathore** [comment]
@blift I checked again the slack docs and found that `chat:write:user` and `chat:write:bot` are really replaced by `chat:write` scope. They just didn't update the [chat.postMessage](https://api.slack.com/methods/chat.postMessage) page. Then I created a slack app to test if it works with `chat:write` scope and found it does. Attaching the screenshot for the reference. So I will be updating the docs and it is not validating against the scope so there should be some other error at your end.

<img width="610" alt="Image" src="https://github.com/user-attachments/assets/99f394df-c5db-4f5f-a5f8-a5e8d44330a6" />
